# DATASET

## Concatenate all downloaded Parquet files by years

In [1]:
import glob
import pandas as pd

In [2]:
def create_df_year(year):
    dfs = []
    # Read parquet files from directory /data/{year} 
    for fin in glob.glob(f'./data/{year}/*.parquet'):
      dfs.append(pd.read_parquet(fin))
    df = pd.concat(dfs)
    return df

def save_as_parquet_compressed(df,name = 'df'):
        return df.to_parquet(f'{name}.parquet.gzip',
              compression='gzip')

In [3]:
df_2022 = create_df_year(2022)

In [4]:
df_2022

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
3399544         2  2022-12-31 23:46:00   2023-01-01 00:11:00              NaN   
3399545         2  2022-12-31 23:13:24   2022-12-31 23:29:08              NaN   
3399546         2  2022-12-31 23:00:49   2022-12-31 23:26:57              NaN   
3399547         1  2022-12-31 23:02:50   2022-12-31 23:16:05              NaN   
3399548         2  2022-12-31 23:00:15   2022-12-31 23:22:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
3399544          11.56         NaN               None            16   
3399545           5.06         NaN               None            75   
3399546          13.35         NaN               None           168   
3399547           0.00         NaN               None           238   
3399548           4.24         NaN               None            85   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
3399544            36             0        39.55    0.0      0.5        8.21   
3399545            50             0        26.23    0.0      0.5        0.00   
3399546           197             0        47.73    0.0      0.5        9.85   
3399547           116             0        12.74    0.0      0.5        0.00   
3399548            37             0        19.60    0.0      0.5        4.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
3399544           0.0                    1.0         49.26   
3399545           0.0                    1.0         30.23   
3399546           0.0                    1.0         59.08   
3399547           0.0                    1.0         16.74   
3399548           0.0                    1.0         25.10   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
3399544                   NaN          NaN  
3399545                   NaN          NaN  
3399546                   NaN          NaN  
3399547                   N

In [5]:
df_2021 = create_df_year(2021)

In [6]:
df_2021

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2021-01-01 00:30:10   2021-01-01 00:36:12              1.0   
1               1  2021-01-01 00:51:20   2021-01-01 00:52:19              1.0   
2               1  2021-01-01 00:43:30   2021-01-01 01:11:06              1.0   
3               1  2021-01-01 00:15:48   2021-01-01 00:31:01              0.0   
4               2  2021-01-01 00:31:49   2021-01-01 00:48:21              1.0   
...           ...                  ...                   ...              ...   
3214364         2  2021-12-31 23:18:02   2021-12-31 23:32:19              NaN   
3214365         2  2021-12-31 23:46:00   2022-01-01 00:07:00              NaN   
3214366         2  2021-12-31 23:46:00   2022-01-01 00:04:00              NaN   
3214367         2  2021-12-31 23:53:00   2022-01-01 00:02:00              NaN   
3214368         2  2021-12-31 23:19:28   2021-12-31 23:26:30              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 2.10         1.0                  N           142   
1                 0.20         1.0                  N           238   
2                14.70         1.0                  N           132   
3                10.60         1.0                  N           138   
4                 4.94         1.0                  N            68   
...                ...         ...                ...           ...   
3214364           3.99         NaN               None           145   
3214365           5.96         NaN               None           236   
3214366           4.99         NaN               None            48   
3214367           1.36         NaN               None            75   
3214368           1.99         NaN               None           170   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                  43             2         8.00    3.0      0.5        0.00   
1                 151             2         3.00    0.5      0.5        0.00   
2                 165             1        42.00    0.5      0.5        8.65   
3                 132             1        29.00    0.5      0.5        6.05   
4                  33             1        16.50    0.5      0.5        4.06   
...               ...           ...          ...    ...      ...         ...   
3214364           107             0        17.42    0.0      0.5        5.80   
3214365           158             0        23.05    0.0      0.5        2.85   
3214366            13             0        20.97    0.0      0.5        5.23   
3214367            75             0        13.20    0.0      0.5        2.00   
3214368           237             0        13.20    0.0      0.5        2.63   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         11.80   
1                 0.0                    0.3          4.30   
2                 0.0                    0.3         51.95   
3                 0.0                    0.3         36.35   
4                 0.0                    0.3         24.36   
...               ...                    ...           ...   
3214364           0.0                    0.3         26.52   
3214365           0.0                    0.3         29.20   
3214366           0.0                    0.3         29.50   
3214367           0.0                    0.3         16.00   
3214368           0.0                    0.3         19.13   

         congestion_surcharge  airport_fee  
0                         2.5          NaN  
1                         0.0          NaN  
2                         0.0          NaN  
3                         0.0          NaN  
4                         2.5          NaN  
...                       ...          ...  
3214364                   NaN          NaN  
3214365                   NaN          NaN  
3214366                   NaN          NaN  
3214367                   N

In [6]:
df_2020 = create_df_year(2020)

In [7]:
df_2020

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1               1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   
2               1  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0   
3               1  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0   
4               2  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0   
...           ...                  ...                   ...              ...   
1461893         2  2020-12-31 23:44:35   2021-01-01 00:01:22              NaN   
1461894         2  2020-12-31 23:41:36   2020-12-31 23:50:32              NaN   
1461895         2  2020-12-31 23:01:17   2020-12-31 23:40:37              NaN   
1461896         2  2020-12-31 23:31:29   2020-12-31 23:44:22              NaN   
1461897         2  2020-12-31 23:12:48   2020-12-31 23:24:51              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 1.20         1.0                  N           238   
1                 1.20         1.0                  N           239   
2                 0.60         1.0                  N           238   
3                 0.80         1.0                  N           238   
4                 0.00         1.0                  N           193   
...                ...         ...                ...           ...   
1461893           9.22         NaN               None           143   
1461894           4.79         NaN               None             4   
1461895          28.00         NaN               None           210   
1461896           7.08         NaN               None           148   
1461897           2.35         NaN               None            18   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 239             1         6.00    3.0      0.5        1.47   
1                 238             1         7.00    3.0      0.5        1.50   
2                 238             1         6.00    3.0      0.5        1.00   
3                 151             1         5.50    0.5      0.5        1.36   
4                 193             2         3.50    0.5      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
1461893            20             0        32.49    0.0      0.5        8.16   
1461894           262             0        13.22    0.0      0.5        3.65   
1461895            78             0        69.31    0.0      0.5        2.75   
1461896           129             0        35.95    0.0      0.5        2.75   
1461897           235             0        17.09    0.0      0.5        2.75   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3         11.27   
1                0.00                    0.3         12.30   
2                0.00                    0.3         10.80   
3                0.00                    0.3          8.16   
4                0.00                    0.3          4.80   
...               ...                    ...           ...   
1461893          0.00                    0.3         43.95   
1461894          0.00                    0.3         20.17   
1461895          6.12                    0.3         78.98   
1461896          0.00                    0.3         39.50   
1461897          0.00                    0.3         20.64   

         congestion_surcharge airport_fee  
0                         2.5        None  
1                         2.5        None  
2                         2.5        None  
3                         0.0        None  
4                         0.0        None  
...                       ...         ...  
1461893                   NaN         NaN  
1461894                   NaN         NaN  
1461895                   NaN         NaN  
1461896                   NaN        

In [8]:
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24649092 entries, 0 to 1461897
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            obj

<hr>

**SUMMARY NO. OF ROWS:** <br>
1. `df_2022` : 39,656,098 rows <br>
2. `df_2021` : 30,904,308 rows <br>
3. `df_2020` : 24,649,092 rows

##  Get the first N rows randomly from each month
`*.parquet.gzip` files are still too large to upload into Github and take too much storage/bandwidth even with Git LFS. (about 1.5GB+)

Hence, reducing the file size is the best solution by filtering the rows for this side project use. Here, we will use first `20,000` rows randomly from each month

### 2022 Dataset CSV and Parquet format

In [2]:
df = pd.read_parquet('2022.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df = df.groupby([df['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime').reset_index(drop=True)
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              2  2002-10-23 16:21:59   2002-10-23 16:22:08              3.0   
1              2  2002-10-27 02:35:45   2002-10-27 03:03:00              1.0   
2              2  2022-01-01 00:06:58   2022-01-01 00:15:52              2.0   
3              2  2022-01-01 00:08:27   2022-01-01 00:21:40              1.0   
4              2  2022-01-01 00:09:02   2022-01-01 00:30:55              1.0   
...          ...                  ...                   ...              ...   
239995         2  2022-12-31 23:46:30   2022-12-31 23:49:45              1.0   
239996         2  2022-12-31 23:53:50   2023-01-01 00:03:14              1.0   
239997         2  2022-12-31 23:56:28   2023-01-01 00:12:48              1.0   
239998         2  2022-12-31 23:56:45   2023-01-01 00:11:34              1.0   
239999         2  2022-12-31 23:58:06   2023-01-01 00:02:08              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                0.00         5.0                  N           217   
1               10.68         1.0                  N           249   
2                0.44         1.0                  N           164   
3                8.95         1.0                  N           138   
4                6.43         1.0                  N           114   
...               ...         ...                ...           ...   
239995           0.53         1.0                  N           143   
239996           0.74         1.0                  N           162   
239997           2.41         1.0                  N           162   
239998           6.95         1.0                  N            43   
239999           0.98         1.0                  N           239   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                217             2         47.0    0.0      0.0        0.00   
1                244             1         32.5    0.5      0.5        7.26   
2                164             1          7.0    0.5      0.5        2.16   
3                116             1         25.0    0.5      0.5        6.82   
4                 24             1         22.0    0.5      0.5        6.45   
...              ...           ...          ...    ...      ...         ...   
239995           143             1          5.8    1.0      0.5        2.70   
239996           162             2          9.3    1.0      0.5        0.00   
239997           143             1         16.3    1.0      0.5        0.00   
239998            88             1         28.9    1.0      0.5        0.00   
239999           238             1          6.5    1.0      0.5       10.00   

        tolls_amount  improvement_surcharge  total_amount  \
0               0.00                    0.3         47.30   
1               0.00                    0.3         43.56   
2               0.00                    0.3         12.96   
3               6.55                    0.3         40.92   
4               0.00                    0.3         32.25   
...              ...                    ...           ...   
239995          0.00                    1.0         13.50   
239996          0.00                    1.0         14.30   
239997          0.00                    1.0         21.30   
239998          0.00                    1.0         33.90   
239999          0.00                    1.0         21.50   

        congestion_surcharge  airport_fee  
0                        0.0         0.00  
1                        2.5         0.00  
2                        2.5         0.00  
3                        0.0         1.25  
4                        2.5         0.00  
...                      ...          ...  
239995                   2.5         0.00  
239996                   2.5         0.00  
239997                   2.5         0.00  
239998                   2.5         0.00  
239999                   2.5         0.0

In [3]:
# Check the count of rows if 10,000 rows were exactly selected or not.
df['tpep_pickup_datetime'].groupby([df['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [4]:
df.to_csv('data_2022.csv', index=False)
df.to_parquet('data_2022.parquet', index=False)

### 2021 Dataset CSV and Parquet format

In [5]:
df1 = pd.read_parquet('2021.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df1 = df1.groupby([df1['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime').reset_index(drop=True)
df1

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              2  2008-12-31 22:59:12   2008-12-31 23:08:12              1.0   
1              2  2009-01-01 00:03:00   2009-01-01 00:07:23              1.0   
2              2  2009-01-01 00:03:10   2009-01-01 17:12:05              1.0   
3              2  2009-01-01 00:04:14   2009-01-01 00:18:10              1.0   
4              2  2009-01-01 00:10:40   2009-01-02 00:06:30              1.0   
...          ...                  ...                   ...              ...   
239995         2  2021-12-31 23:39:29   2021-12-31 23:47:58              1.0   
239996         2  2021-12-31 23:48:24   2022-01-01 00:07:40              1.0   
239997         2  2021-12-31 23:49:06   2021-12-31 23:50:21              1.0   
239998         2  2021-12-31 23:50:15   2021-12-31 23:58:52              1.0   
239999         2  2022-01-17 12:51:59   2022-01-17 12:59:05              4.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                1.27         1.0                  N           113   
1                1.04         1.0                  N           163   
2                0.97         1.0                  N           163   
3                1.35         1.0                  N           234   
4                1.06         1.0                  N           162   
...               ...         ...                ...           ...   
239995           1.78         1.0                  N            43   
239996           3.62         1.0                  N           161   
239997           0.46         1.0                  N           238   
239998           1.47         1.0                  N           141   
239999           0.88         1.0                  N            50   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                234             2          7.5    0.0      0.5        0.00   
1                229             2          6.0    0.0      0.5        0.00   
2                163             1          7.5    0.0      0.5        2.16   
3                158             2         10.0    0.0      0.5        0.00   
4                237             2          6.5    0.0      0.5        0.00   
...              ...           ...          ...    ...      ...         ...   
239995           263             1          8.5    0.5      0.5        2.46   
239996           231             1         15.5    0.5      0.5        3.86   
239997           238             1          3.5    0.5      0.5        1.82   
239998           142             1          8.0    0.5      0.5        2.36   
239999           161             1          6.5    0.5      0.5        2.06   

        tolls_amount  improvement_surcharge  total_amount  \
0                0.0                    0.3         10.80   
1                0.0                    0.3          9.30   
2                0.0                    0.3         12.96   
3                0.0                    0.3         13.30   
4                0.0                    0.3          9.80   
...              ...                    ...           ...   
239995           0.0                    0.3         14.76   
239996           0.0                    0.3         23.16   
239997           0.0                    0.3          9.12   
239998           0.0                    0.3         14.16   
239999           0.0                    0.3         12.36   

        congestion_surcharge  airport_fee  
0                        2.5          0.0  
1                        2.5          0.0  
2                        2.5          0.0  
3                        2.5          0.0  
4                        2.5          0.0  
...                      ...          ...  
239995                   2.5          0.0  
239996                   2.5          0.0  
239997                   2.5          0.0  
239998                   2.5          0.0  
239999                   2.5          0.

In [6]:
# Check the count of rows if 10,000 rows were exactly  selected or not.
df1['tpep_pickup_datetime'].groupby([df1['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [7]:
df1.to_csv('data_2021.csv', index=False)
df1.to_parquet('data_2021.parquet', index=False)

### 2020 Dataset CSV and Parquet format

In [8]:
df2 = pd.read_parquet('2020.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df2 = df2.groupby([df2['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime').reset_index(drop=True)
df2

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              2  2008-12-31 23:05:27   2009-01-01 12:13:35              1.0   
1              2  2009-01-01 23:51:06   2009-01-02 00:02:23              2.0   
2              2  2019-12-31 18:21:19   2020-01-01 17:57:53              1.0   
3              2  2019-12-31 19:13:46   2019-12-31 19:23:49              1.0   
4              2  2019-12-31 23:59:33   2020-01-01 00:03:25              2.0   
...          ...                  ...                   ...              ...   
239995         1  2020-12-31 23:42:21   2021-01-01 00:01:51              1.0   
239996         2  2020-12-31 23:42:59   2020-12-31 23:54:22              1.0   
239997         2  2020-12-31 23:45:42   2020-12-31 23:56:25              2.0   
239998         1  2020-12-31 23:50:34   2020-12-31 23:53:46              1.0   
239999         1  2020-12-31 23:52:35   2021-01-01 00:12:35              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                0.00         1.0                  N           193   
1                0.00         1.0                  N           193   
2               17.86         2.0                  N           132   
3                1.26         1.0                  N           231   
4                0.78         1.0                  N           237   
...               ...         ...                ...           ...   
239995           7.70         1.0                  N           230   
239996           2.41         1.0                  N            79   
239997           0.67         1.0                  N           234   
239998           0.80         1.0                  N            50   
239999           2.80         1.0                  N           246   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                193             1          0.0    0.0      0.0        0.00   
1                193             2          7.5    0.0      0.5        0.00   
2                163             1         52.0    0.0      0.5        5.00   
3                148             2          8.0    0.5      0.5        0.00   
4                237             1          5.0    0.5      0.5        1.00   
...              ...           ...          ...    ...      ...         ...   
239995            70             2         23.5    3.0      0.5        0.00   
239996            88             2         11.0    0.5      0.5        0.00   
239997           164             1          8.0    0.5      0.5        1.00   
239998           143             1          4.5    3.0      0.5        1.65   
239999           162             1         15.0    3.0      0.5        3.75   

        tolls_amount  improvement_surcharge  total_amount  \
0                0.0                    0.0          0.00   
1                0.0                    0.3          8.30   
2                0.0                    0.3         57.80   
3                0.0                    0.3         11.80   
4                0.0                    0.3          9.80   
...              ...                    ...           ...   
239995           0.0                    0.3         27.30   
239996           0.0                    0.3         14.80   
239997           0.0                    0.3         12.80   
239998           0.0                    0.3          9.95   
239999           0.0                    0.3         22.55   

        congestion_surcharge  airport_fee  
0                        0.0          NaN  
1                        0.0          NaN  
2                        0.0          NaN  
3                        2.5          NaN  
4                        2.5          NaN  
...                      ...          ...  
239995                   2.5          NaN  
239996                   2.5          NaN  
239997                   2.5          NaN  
239998                   2.5          NaN  
239999                   2.5          Na

In [9]:
# Check the count of rows if 10,000 rows were exactly  selected or not.
df2['tpep_pickup_datetime'].groupby([df2['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [10]:
df2.to_csv('data_2020.csv', index=False)
df2.to_parquet('data_2020.parquet', index=False)

<hr>

**SUMMARY CSV FILE SIZE:** <br>
1. `data_2022.csv` : 24 MB <br>
2. `data_2021.csv` : 24 MB <br>
3. `data_2020.csv` : 23 MB<br>

**SUMMARY PARQUET FILE SIZE:** <br>
1. `data_2022.parquet` : 6 MB <br>
2. `data_2021.parquet` : 5 MB <br>
3. `data_2020.parquet` : 5 MB

## Miscs

In [13]:
#Reading data using pandas read_parquet()
pd.read_parquet('data_2020.parquet')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              2  2008-12-31 23:05:27   2009-01-01 12:13:35              1.0   
1              2  2009-01-01 23:51:06   2009-01-02 00:02:23              2.0   
2              2  2019-12-31 18:21:19   2020-01-01 17:57:53              1.0   
3              2  2019-12-31 19:13:46   2019-12-31 19:23:49              1.0   
4              2  2019-12-31 23:59:33   2020-01-01 00:03:25              2.0   
...          ...                  ...                   ...              ...   
239995         1  2020-12-31 23:42:21   2021-01-01 00:01:51              1.0   
239996         2  2020-12-31 23:42:59   2020-12-31 23:54:22              1.0   
239997         2  2020-12-31 23:45:42   2020-12-31 23:56:25              2.0   
239998         1  2020-12-31 23:50:34   2020-12-31 23:53:46              1.0   
239999         1  2020-12-31 23:52:35   2021-01-01 00:12:35              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                0.00         1.0                  N           193   
1                0.00         1.0                  N           193   
2               17.86         2.0                  N           132   
3                1.26         1.0                  N           231   
4                0.78         1.0                  N           237   
...               ...         ...                ...           ...   
239995           7.70         1.0                  N           230   
239996           2.41         1.0                  N            79   
239997           0.67         1.0                  N           234   
239998           0.80         1.0                  N            50   
239999           2.80         1.0                  N           246   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                193             1          0.0    0.0      0.0        0.00   
1                193             2          7.5    0.0      0.5        0.00   
2                163             1         52.0    0.0      0.5        5.00   
3                148             2          8.0    0.5      0.5        0.00   
4                237             1          5.0    0.5      0.5        1.00   
...              ...           ...          ...    ...      ...         ...   
239995            70             2         23.5    3.0      0.5        0.00   
239996            88             2         11.0    0.5      0.5        0.00   
239997           164             1          8.0    0.5      0.5        1.00   
239998           143             1          4.5    3.0      0.5        1.65   
239999           162             1         15.0    3.0      0.5        3.75   

        tolls_amount  improvement_surcharge  total_amount  \
0                0.0                    0.0          0.00   
1                0.0                    0.3          8.30   
2                0.0                    0.3         57.80   
3                0.0                    0.3         11.80   
4                0.0                    0.3          9.80   
...              ...                    ...           ...   
239995           0.0                    0.3         27.30   
239996           0.0                    0.3         14.80   
239997           0.0                    0.3         12.80   
239998           0.0                    0.3          9.95   
239999           0.0                    0.3         22.55   

        congestion_surcharge  airport_fee  
0                        0.0          NaN  
1                        0.0          NaN  
2                        0.0          NaN  
3                        2.5          NaN  
4                        2.5          NaN  
...                      ...          ...  
239995                   2.5          NaN  
239996                   2.5          NaN  
239997                   2.5          NaN  
239998                   2.5          NaN  
239999                   2.5          Na

In [ ]:
# 2019 and 2018 are not used for this project
# df_2019 = create_df_year(2019)
# df_2018 = create_df_year(2018)